contracts/HATVaultsRegistry.sol

```diff
@@ -0,0 +1,563 @@
+    }
+
+    /**
+    * @notice change the UI visibility of a vault
+    * only calleable by the owner of the contract
+    * @param _vault the vault to update
+    * @param _visible is this vault visible in the UI
+    * This parameter can be used by the UI to include or exclude the vault
+    */
+    function setVaultVisibility(address _vault, bool _visible) external onlyOwner {
+        emit SetVaultVisibility(_vault, _visible);
+    }
+
+    /**
+    * @notice change the description of a vault
+    * only calleable by the owner of the contract
+    * @param _vault the vault to update
+    * @param _descriptionHash the hash of the vault's description.
+    */
+    function setVaultDescription(address _vault, string memory _descriptionHash) external onlyOwner {
+        emit SetVaultDescription(_vault, _descriptionHash);
+    }
+
+    /**
+    * @notice Transfer the part of the bounty that is supposed to be swapped
+    * into HAT tokens from the HATVault to the registry, and keep track of the
+    * amounts to be swapped and sent/burnt in a later transaction
+    * @param _asset The vault's native token
+    * @param _hacker The address of the beneficiary of the bounty
+    * @param _hackersHatReward The amount of the vault's native token to be
+    * swapped to HAT tokens and sent to the hacker via a vesting contract
+    * @param _governanceHatReward The amount of the vault's native token to be
+    * swapped to HAT tokens and sent to governance
+    */
+    function addTokensToSwap(
+        IERC20 _asset,
+        address _hacker,
+        uint256 _hackersHatReward,
+        uint256 _governanceHatReward
+    ) external {
+        uint256 amount = _hackersHatReward + _governanceHatReward;
+        hackersHatReward[address(_asset)][_hacker] += _hackersHatReward;
+        governanceHatReward[address(_asset)] += _governanceHatReward;
+        _asset.safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    /**
+    * @notice Called by governance to swap vault's native tokens to HAT tokens
+    * and distribute the HAT tokens: Send to governance their share and send to
+    * beneficiaries their share through a vesting contract.
+    * @param _asset The vault's native token
+    * @param _beneficiaries Addresses of beneficiaries
+    * @param _amountOutMinimum Minimum amount of HAT tokens at swap
+    * @param _routingContract Routing contract to call for the swap
+    * @param _routingPayload Payload to send to the _routingContract for the
+    * swap
+    */
+    function swapAndSend(
+        address _asset,
+        address[] calldata _beneficiaries,
+        uint256 _amountOutMinimum,
+        address _routingContract,
+        bytes calldata _routingPayload
+    ) external onlyOwner {
+        // Needed to avoid a stack too deep error
+        SwapData memory swapData;
+        for (uint256 i = 0; i < _beneficiaries.length; i++) {
+            swapData.totalHackersHatReward += hackersHatReward[_asset][_beneficiaries[i]];
```

Comment:
The caller of this function needs to know the total amount to be swapped to be able to specify a reasonable value for amountOutMininum. This is not so easy now. Should we add a getter for hackersHatReward, so it is easy to read from the outside?

Reply:
we have all the info in the approveClaim event so we can read from that. maybe we should have addTokensToSwap emit an event also?

Reply:
then maybe would be better to specify the minimum ratio instead of the minimum amount and then calculate the minimum amount in the function

Reply:
hackersHatReward is public so not sure I understand the issue

Reply:
i think anyway to specify minimum ratio makes more sense, no?

Reply:
> hackersHatReward is public so not sure I understand the issue
nevermind you right


> i think anyway to specify minimum ratio makes more sense, no?

No it does not. The price depends on the amount (if you sell 1 ether for 2000/ETH does not mean you can sell 1B ether for 2000/ETH). So setting the mimim price we want to get for the entire batch is the logical thing. Specifying the average price of the batch is not better UX.

Reply:
ok gotcha :) i actually thought i does mean that.....